### PR TITLE
Factor analysis minor improvements

### PR DESCRIPTION
**Pull request by DominiqueMakowski**

### CODE REVIEW

File: R/principal_components.R [hunk 1]
```diff
@@ -420,10 +420,6 @@ principal_components.data.frame <- function(x,
   } else if (n >= ncol(x)) {
     n <- ncol(x) - 1
   }
-  # sanity check - we need at least two factors
```
Comment (DominiqueMakowski) [inline R/principal_components.R:423]: This was making that whenever we ran `factor_analysis(mtcars, n=1)` it stubbornely returned 2 factors...

File: R/n_factors.R [hunk 2]
```diff
@@ -138,6 +138,13 @@ n_factors <- function(x,
   }
 
   # Get only numeric
+  if(!all(vapply(x, is.numeric, TRUE))) {
+    warning(paste0(
```
Comment (strengejacke) [inline R/n_factors.R:142]: Use `insight::format_warning()` here

File: R/cluster_analysis.R [hunk 3]
```diff
@@ -449,6 +449,12 @@ visualisation_recipe.cluster_analysis_summary <- function(x, ...) {
 #' @export
 visualisation_recipe.cluster_analysis <- function(x, show_data = "text", ...) {
   ori_data <- stats::na.omit(attributes(x)$data)
+
+  # Check number of columns: if more than 2, display PCs, if less, fail
+  if(ncol(ori_data) <= 2) {
+    stop("Less than 2 variables in the dataset. Cannot compute enough principal components to represent clustering.")
+  }
```
Comment (strengejacke) [inline R/cluster_analysis.R:456]: Use `insight::format_error()` here

**Review by strengejacke — CHANGES_REQUESTED** [review]

**Comment by strengejacke** [comment]
Any chance we can also have tests for this functionality? :-)

File: R/convert_efa_to_cfa.R [hunk 4]
```diff
@@ -85,9 +96,20 @@ efa_to_cfa <- convert_efa_to_cfa
...
+
+    # Find correct subset
+    items <- loadings[loadings$Component == unique(loadings$Component)[i], ]
+
+    # Find corresponding items
+    items <- as.character(loadings[loadings$Component == unique(loadings$Component)[i], "Variable"])
+
+    # Subset if need be to keep only a certain number
+    if(!is.null(max_per_dimension) && max_per_dimension > 0) items <- as.character(na.omit(items[1:max_per_dimension]))
+
```
Comment (strengejacke) [inline R/convert_efa_to_cfa.R:108]: Use `stats::na.omit()` here.

File: R/n_factors.R [hunk 5]
```diff
@@ -564,6 +571,31 @@ print.n_clusters <- print.n_factors
...
+    fa <- "fa"
+  } else {
+    fa <- "pc"
+  }
+
+  out <- psych::fa.parallel(cor, n.obs = nobs, fa=fa, plot=FALSE, fm="ml")
+
+
+  .data_frame(
+    n_Factors = as.numeric(na.omit(c(out$nfact, out$ncomp))),
```
Comment (strengejacke) [inline R/n_factors.R:593]: Use `stats::na.omit()` here.

File: R/n_factors.R [hunk 6]
```diff
@@ -564,6 +571,31 @@ print.n_clusters <- print.n_factors
...
+  # Altnerative version of parralel analysis
+  # Not used because already included in nFactors
+
+  if (tolower(type) %in% c("fa", "factor", "efa")) {
+    fa <- "fa"
+  } else {
+    fa <- "pc"
+  }
+
+  out <- psych::fa.parallel(cor, n.obs = nobs, fa=fa, plot=FALSE, fm="ml")
```
Comment (strengejacke) [inline R/n_factors.R:589]: Spaces around `=`.

**Review by strengejacke — CHANGES_REQUESTED** [review]

File: R/convert_efa_to_cfa.R [hunk 7]
```diff
@@ -33,8 +36,15 @@ convert_efa_to_cfa <- function(model, ...) {
 #' @rdname convert_efa_to_cfa
 #' @inheritParams model_parameters.principal
 #' @export
-convert_efa_to_cfa.fa <- function(model, threshold = "max", names = NULL, ...) {
-  .efa_to_cfa(model_parameters(model, threshold = threshold, ...), names = names, ...)
+convert_efa_to_cfa.fa <- function(model,
+                                  threshold = "max",
+                                  names = NULL,
+                                  max_per_dimension = NULL,
```
Comment (strengejacke) [inline R/convert_efa_to_cfa.R:42]: `max_per_dimension` needs to be documented,

**Review by strengejacke — CHANGES_REQUESTED** [review]

File: R/convert_efa_to_cfa.R [hunk 8]
```diff
@@ -85,9 +96,20 @@ efa_to_cfa <- convert_efa_to_cfa
...
   for (i in seq_along(names)) {
+
+    # Find correct subset
+    items <- loadings[loadings$Component == unique(loadings$Component)[i], ]
+
+    # Find corresponding items
+    items <- as.character(loadings[loadings$Component == unique(loadings$Component)[i], "Variable"])
+
+    # Subset if need be to keep only a certain number
+    if(!is.null(max_per_dimension) && max_per_dimension > 0) items <- as.character(na.omit(items[1:max_per_dimension]))
```
Comment (strengejacke) [inline R/convert_efa_to_cfa.R:107]: > Place a space before left parenthesis, except in a function call.

File: R/cluster_analysis.R [hunk 9]
```diff
@@ -449,6 +449,12 @@ visualisation_recipe.cluster_analysis_summary <- function(x, ...) {
 #' @export
 visualisation_recipe.cluster_analysis <- function(x, show_data = "text", ...) {
   ori_data <- stats::na.omit(attributes(x)$data)
+
+  # Check number of columns: if more than 2, display PCs, if less, fail
+  if(ncol(ori_data) <= 2) {
```
Comment (strengejacke) [inline R/cluster_analysis.R:454]: > Place a space before left parenthesis, except in a function call.

File: R/n_factors.R [hunk 10]
```diff
@@ -138,6 +138,13 @@ n_factors <- function(x,
   }
 
   # Get only numeric
+  if(!all(vapply(x, is.numeric, TRUE))) {
```
Comment (strengejacke) [inline R/n_factors.R:141]: > Place a space before left parenthesis, except in a function call.

File: R/n_factors.R [hunk 11]
```diff
@@ -138,6 +138,13 @@ n_factors <- function(x,
...
+    warning(paste0(
+      "Some variables are not numeric (",
+      paste0(names(x)[!vapply(x, is.numeric, TRUE)], collapse = ", "),
```
Comment (strengejacke) [inline R/n_factors.R:144]: > toString(.) is more expressive than paste(., collapse = ", ")

**Review by strengejacke — CHANGES_REQUESTED** [review]

**Comment by strengejacke** [comment]
Furthermore:
- [ ] RD files need to be updated
- [ ] Tests added

**Comment by strengejacke** [comment]
_bump_

**Comment by strengejacke** [comment]
Can this be merged?

**Comment by strengejacke** [comment]
@DominiqueMakowski can this be merged?

**Comment by DominiqueMakowski** [comment]
Oh wow completely missed that. Yes I think it's good

**Comment by strengejacke** [comment]
Let me see if checks pass, and then I'll merge.

**Pull request merged**